### PR TITLE
[None][feat] Consolidate the DSpark draft paths and support standalone drafters

### PR DESCRIPTION
**Pull request by dc3671**
## Description

"DSpark" names two unrelated things in this repo, and `decoding_type: DSpark` only reaches one of them:

- **DSv4-Pro** — draft lives inside the target checkpoint under `mtp.*`, three full DeepSeek-V4 blocks.
- **Kimi K3** — draft is a standalone checkpoint. It has to be configured as `decoding_type: DFlash`, because that is where the DSpark semantics landed in #17327.

The difference is packaging, not algorithm. `is_dspark` in `modeling_speculative.py` is already true exactly when the Markov head, confidence head, or `shift_label` convention is present — i.e. **DFlash is DSpark with those heads off**, not the reverse. This PR makes the API match that.

```mermaid
flowchart TD
    F["decoding_type: DFlash"] --> D["DFlashForCausalLM<br/>DFlashWorker"]
    C["decoding_type: DSpark"] --> Q{"draft weights inside<br/>the target checkpoint?"}
    Q -->|yes| V["DSv4DSparkForCausalLM<br/>DSv4DSparkWorker"]
    Q -->|no| K["GQADSparkForCausalLM<br/>DSparkWorker"]
```

Both dispatches — draft model and worker/metadata — read the same resolved-once flag, so they cannot drift apart.

Getting there needed the dependency inverted first. `get_draft_model` had to import every concrete draft model, which forced a lazy import to break the `modeling_dspark → modeling_deepseekv4 → modeling_speculative` cycle. It is now a builder registry with a string index in `_arch_index.py`, carrying over the three rules `register_auto_model` already encodes. With the cycle gone, `DFlashForCausalLM` moves to its own `modeling_dflash.py`, the two duplicate Markov implementations merge into one kernel, and `models/dspark/` folds into `modeling_dspark.py`.

Standalone drafters also load as published now. Both public K3 checkpoints put `markov_rank` / `enable_confidence_head` at the top level of `config.json` and name the head tensors after the modules that own them (`markov_head.markov_w1.weight`), while the reader understood only a nested `dflash_config` and bare tensor names — so the Markov and confidence heads were dropped in silence, costing acceptance with nothing raised. One resolver now covers every spelling, shared with `validate_speculative_config` so the model and the user-visible config cannot disagree, and shipping head weights that resolve to rank 0 is an error rather than a silent drop. `shift_label` defaults on for a DSpark drafter: both checkpoints set `block_size == max_draft_len`, where the DFlash slot layout runs one slot past the block and reads the next request's anchor.

One target-side fix rides along, because it is what makes the standalone drafter worth running. K3 folds the residual into a running prefix sum, and the DSpark capture was taking that raw value. The drafter is distilled on the aggregated stream — the pre-norm softmax mixture its next consumer sees — which layer i+1 already computes, so the tap now rides along at no extra cost. Ground truth is SGLang `kimi_k3.py:2589 _dspark_capture_stream`, whose `attn_res is None` fallback is exactly what we were capturing. Worth 4.5pt of draft acceptance on a separate harness (AR 66.9% → 71.4%).

**Two intentional behaviour changes:**

- `decoding_type: DFlash` now rejects a drafter that declares the DSpark head set, with a message pointing at `DSpark`.
- `moe_load_balancer` is propagated only to an embedded draft — what `external_drafter_config_kwargs` already documented.

Commits are sequenced so each is separately reviewable; the pure moves and the rename are behaviour-preserving.

## Test Coverage

Nine tests, no new files — each goes in the file that already owns its subject and reuses that file's fixtures:

- `test_kimi_k3_dspark_semantics.py` — the published config and weight spellings activating the heads, head weights with an unresolvable rank, the DFlash refusal, the GQA precondition.
- `test_dspark_worker.py` — form-based worker and draft-KV routing, and the worker policies coming off the drafter rather than hardcoded.
- `test_dspark_eplb_config.py` — the deployment-form probe across five checkpoint layouts.
- `tests/unittest/others/test_lazy_model_zoo.py` — spec-mode index drift against the decorators.

All but the last guard silent degradation: a dropped Markov head lowers acceptance without failing anything, which is what the accuracy column above cannot see.

GSM8K, 16×GB200, 1319 questions, greedy, RadixArk K3 DSpark drafter:

| | Accuracy | Acceptance length |
|---|---|---|
| Before, `decoding_type: DFlash`, pre-converted drafter | 96.59 | 3.798 |
| After, `decoding_type: DSpark`, pre-converted drafter | 96.66 | 3.814 |
| `DSpark`, drafter as published, raw prefix-sum capture | 97.04 | 3.810 |
| `DSpark`, drafter as published, aggregated-stream capture | 95.98 | 4.156 |
| the same run repeated, nothing changed | 96.44 | 4.158 |

Row 3 → row 4 is the capture fix, on one variable. Acceptance length is reproducible to 0.002 across the repeat; accuracy spans 0.46 between two identical runs, which matches lm-eval's own ±0.4666 stderr, so no accuracy claim is made in either direction. The published-drafter rows need no conversion step, which is the point of the spelling work — the converter that produced our original checkpoint no longer exists.

The refactoring commits were also checked beyond the suite: the Markov merge by bit-for-bit comparison against both pre-merge implementations, the package fold by an AST symbol diff, the `DSv4` rename by mechanically rewriting the added lines back and matching them against the removed ones.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- Consolidates DSpark implementation into `modeling_dspark.py`.
- Moves DFlash implementation into `modeling_dflash.py`.
- Adds lazy draft-model builder registration through `DRAFT_MODEL_BUILDER_MAPPING` and `SPEC_MODE_TO_MODULE`.
- Supports embedded DeepSeek-V4-Pro and standalone Kimi K3 DSpark drafts.
- Routes models, workers, metadata, KV-cache policy, and EPLB propagation by deployment type.
- Adds published K3 configuration and tensor-name compatibility.
- Captures the aggregated K3 stream consumed by the drafter.
- Adds DSpark attention-backend configuration and validation.
- Rejects DSpark head settings when `decoding_type` is `DFlash`.
- Enables default `shift_label` behavior for DSpark drafters.
- Adds configurable `num_fewshot` support to `LmEvalEvaluator`.

Review focus:

- Verify builder registration, checkpoint classification, and deployment routing.
- Verify published K3 tensor aliases and GQA validation.
- Verify CUDA-graph paths, captured-stream lifetimes, and rolling KV-window behavior.
- Verify configuration values and test-list syntax against repository guidelines.

## QA Engineer Review

- Added `TestQwen3_8B.test_dspark`.
- Added `test_spec_mode_index_matches_decorators()`.
- Added `test_capture_taps_the_next_layers_aggregated_stream()`.
- Updated DSpark attention, CUDA-graph, draft, head, EPLB, worker, and Kimi K3 semantic tests.
- Added Qwen3-8B DSpark entries to `tests/integration/test_lists/test-db/l0_h100.yml` and `tests/integration/test_lists/test-db/l0_b200.yml`.
- Added acceptance-length and GSM8K references for the integration test.
- The integration tests are covered by the H100 and B200 pre-merge test lists.
- Verdict: sufficient.

### CODE REVIEW

**Comment by dc3671** [comment]
Related to https://github.com/NVIDIA/TensorRT-LLM/pull/16813 @chungen04 for vis.

**Review by tburt-nv — APPROVED** [review]

**Review by zhaoyuanh-nvidia — APPROVED** [review]

File: tests/integration/defs/accuracy/references/gsm8k.yaml [hunk 1]
```diff
@@ -190,6 +190,14 @@ Qwen3/Qwen3-8B:
     accuracy: 87.1114
   - spec_dec_algo: DFlash
     accuracy: 87.1114
+  # 0-shot chat, for the acceptance-length gate only. The model stops emitting
```
Comment (Wanli-Jiang) [inline tests/integration/defs/accuracy/references/gsm8k.yaml]: if so, could we just skip the accuracy checking and only check accept length?

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 2]
```diff
@@ -4856,6 +4880,61 @@ def test_dflash(self):
...
+                 **pytorch_config,
+                 kv_cache_config=kv_cache_config,
+                 max_stats_len=-1,
+                 enable_iter_perf_stats=True,
+                 speculative_config=spec_config) as llm:
+            task = GSM8K(self.MODEL_NAME)
+            task.evaluate(llm,
+                          extra_acc_spec="zero_shot_al_only",
+                          extra_evaluator_kwargs=dict(
+                              num_fewshot=0,
```
Comment (Wanli-Jiang) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py:5121]: better to add comments about why set 0 shot here.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 3]
```diff
@@ -1630,6 +1638,8 @@ def forward(
                 self.self_attention_res_proj,
                 self.self_attention_res_norm,
             )
+        if capture is not None:
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py:1710]: This capture taps layer i's aggregated stream (layer i-1's), not the raw prefix_sum — that's the whole 4.5pt fix, but it fails silently. Any refactor that taps prefix_sum instead just quietly drops acceptance, nothing errors. And nothing tests it: test_dflash_spec_metadata_capture_prefix_sum_convention (tests/unittest/_torch/speculative/hw_agnostic/test_kimi_k3_dflash_scaffold.py:317) only checks buffer routing, and test_kimi_linear_model_threads_spec_metadata (tests/unittest/_torch/speculative/hw_agnostic/test_kimi_k3_dflash_scaffold.py:302) is just a source-string check. Both would still pass if capture reverted to prefix_sum. Can we add a regression test that asserts the captured tensor for layer j matches j+1's aggregated stream, not layer j's prefix_sum? Worth covering the last-layer branch at tensorrt_llm/_torch/models/modeling_kimi_linear.py:1789-1802 too — it uses different weights and is untested.

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py [hunk 4]
```diff
@@ -1744,19 +1754,52 @@ def forward(
...
-                # DFlash hidden-state capture. K3's attn-residual scheme
-                # already folds the residual into the running prefix sum
-                # returned by each layer, so unlike Qwen3/Llama we pass the
-                # full hidden state with residual=None. Whether the drafter
-                # is trained against this prefix sum or some other tap point
-                # must be confirmed against the K3 drafter training recipe
-                # before real weights are used.
-                spec_metadata.maybe_capture_hidden_states(layer.layer_idx, hidden_states, None)
+
+        # The last layer has no successor, so this one recompute is
```
Comment (Wanli-Jiang) [inline tensorrt_llm/_torch/models/modeling_kimi_linear.py:1853]: better to construct a unittest to cover this no-successor part.

**Review by Wanli-Jiang — APPROVED** [review]
LGTM from kimi-k3 modeling viewpoint.

File: tests/integration/defs/accuracy/references/gsm8k.yaml (hunk 1 above)
Comment (dc3671) [reply tests/integration/defs/accuracy/references/gsm8k.yaml]: we need to run gsm8k test first to so we can get AL from `llm.get_stats` in `_compute_acceptance_length`

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 4 above)
Comment (dc3671) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py:1853]: done

File: tensorrt_llm/_torch/models/modeling_kimi_linear.py (hunk 3 above)
Comment (dc3671) [reply tensorrt_llm/_torch/models/modeling_kimi_linear.py:1710]: done. thanks for this mention!

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 2 above)
Comment (dc3671) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py:5121]: done

**Review by litaotju — APPROVED** [review]

**Comment by dc3671** [comment]
Seem the **multi-gpu test is not needed**. I asked agent to search for dflash/dspark related tests. Only DSv4DSpark needs multi gpu, but is waived now(coming back in https://github.com/NVIDIA/TensorRT-LLM/pull/18182). Other dflash tests passed in single-GPU tests. And I have tested locally, so I will just merge this PR for now.
